docs/design/2024-09-06-maintain-priority-queue-in-memory.md

```diff
@@ -0,0 +1,557 @@
+| Analysis Interval    | Time since the last analysis execution for the table. The bigger interval should have a higher priority than the smaller interval.                                                            | Applying a logarithmic transformation, namely log10(1 + Analysis Interval). To further compress the rate of growth for larger values, we can consider taking the logarithmic square root of x. The final formula is log10(1 + √Analysis Interval). [Check the graph.](https://www.desmos.com/calculator/plhtkfqhx9) | Stats Cache                  |
+| Special Event        | For example, the table has a new index but it hasn't been analyzed yet.                                                                                                                       | HasNewIndexWithoutStats: 2                                                                                                                                                                                                                                                                                          | Table Info (Scan all tables) |
+
+Currently, we rebuild the priority queue every 3 seconds by scanning all schemas and tables.
+For count and modify_count, the priority queue gets them from the statistics cache.
+
+## Detailed Design
+
+### Goals
+
+1. Avoid scanning the schema cache every time.
+2. Reflect the DDL changes to the priority queue.
+3. Ensure that DDL events can be received with the correct reliability and order.
+
+### Update the priority queue based on DML and DDL changes
+
+As you can see from the table above, the information we need is divided into two categories, the changes to the table after the DML occurs, and the DDL changes to the table.
+For DML, we can get it from the stats cache as we load all table's modify_count and count by default, so we can just recalculate the score when table data changes.
+For DDL, we need a reliable way to get the DDL events that have occurred and recalculate the score as a result of the change in the table structure.
+
+### Incremental Update
+
+#### DML
+
+Currently, the stats module maintains the most basic information for all tables (modify_count and count) in the stats cache, and we can maintain the priority queue along the lines of incrementally updating it.
+
+1. We start a background worker to call the statsHandle.Update(do.InfoSchema()) function every 3 seconds(stats lease).
+2. In the stats cache struct, we maintain an atomic timestamp called: maxTblStatsVer
```

Comment:
maybe mention what's the definition of `maxTblStatsVer` here.

Reply:
I have added a link for it.